sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtilsSuite.scala

```diff
@@ -65,4 +67,60 @@ class JdbcUtilsSuite extends SparkFunSuite {
     }
     assert(mismatchedInput.getMessage.contains("mismatched input '.' expecting"))
   }
+
+  test("Custom ConnectionFactory falls back to default when no factory specified") {
+    val options = new JDBCOptions(Map[String, String](
+      "url" -> "jdbc:mysql://foo.com/bar",
+      "dbtable" -> "table")
+    )
+    assert(options.connectionFactoryProvider eq DefaultConnectionFactoryProvider)
+  }
+
+  test("JdbcUtils uses the specified connection factory") {
+    val options = new JDBCOptions(Map[String, String](
+      "url" -> "jdbc:mysql://foo.com/bar",
+      "dbtable" -> "table",
+      JDBCOptions.JDBC_CONNECTION_FACTORY_PROVIDER ->
+        "org.apache.spark.sql.execution.datasources.jdbc.TestFactory")
+    )
+    val x = intercept[RuntimeException] {
+      JdbcUtils.createConnectionFactory(options)()
+    }
+    assert(x.getMessage == "This message will be tested in test")
+  }
+
+  test("invalid connection factory throws IllegalArgumentException") {
+
+    val nonexisting = intercept[IllegalArgumentException] {
+      new JDBCOptions(Map[String, String](
+        "url" -> "jdbc:mysql://foo.com/bar",
+        "dbtable" -> "table",
+        JDBCOptions.JDBC_CONNECTION_FACTORY_PROVIDER -> "notexistingclass")
+      )
+    }
+    assert(nonexisting.getMessage == "notexistingclass is not a valid ConnectionFactoryProvider")
+
+    val missingTrait = intercept[IllegalArgumentException] {
+      new JDBCOptions(Map[String, String](
+        "url" -> "jdbc:mysql://foo.com/bar",
+        "dbtable" -> "table",
+        JDBCOptions.JDBC_CONNECTION_FACTORY_PROVIDER ->
+          "org.apache.spark.sql.execution.datasources.jdbc.BadFactory")
+      )
+    }
+    assert(missingTrait.getMessage ==
+      "org.apache.spark.sql.execution.datasources.jdbc.BadFactory" +
+        " is not a valid ConnectionFactoryProvider")
+
+  }
+}
+
+// this one does not implement ConnectionFactoryProvider
+class BadFactory {
+
+}
+
+class TestFactory extends ConnectionFactoryProvider {
+  override def createConnectionFactory(options: JDBCOptions): () => Connection =
```

Comment:
> Without these changes we had to copy most of the spark jdbc package into our own codebase
to allow us to create our own connection factory

Trying to understand. So here we are passing JDBCOptions which is a spark class. I thought the intention was to make the custom connection factory code to not depend on spark classes ?

Reply:
The intent is for the spark jdbc package to allow for custom connection factories by specifying a classname in the options to spark.sql.read. This class has to implement the trait ConnectionFactoryProvider which in this new setup would also be provided by spark jdbc. The only reason we had to copy code into our code base was to make changes to JdbcUtils to allow this and for those classes in spark jdbc that use JdbcUtils.createConnectionFactory to use ours instead. This PR would allow for this behavior out of the box but use the default factory when no className for a ConnectionFactoryProvider is specified.

Reply:
I see. OK.